src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -190,38 +190,63 @@ public static void waitForServicesToStart(Collection<GreengrassService> services
             throws InterruptedException, ServiceUpdateException {
         // Relying on the fact that all service lifecycle steps should have timeouts,
         // assuming this loop will not get stuck waiting forever
-        while (true) {
-            boolean allServicesRunning = true;
-            for (GreengrassService service : servicesToTrack) {
-                State state = service.getState();
-
-                // If a service is previously BROKEN, its state might have not been updated yet when this check
-                // executes. Therefore we first check the service state has been updated since merge map occurs.
-                if (service.getStateModTime() > mergeTime && State.BROKEN.equals(state)) {
-                    logger.atWarn(MERGE_CONFIG_EVENT_KEY).kv(SERVICE_NAME_LOG_KEY, service.getName())
-                            .log("merge-config-service BROKEN");
-                    throw new ServiceUpdateException(
-                            String.format("Service %s in broken state after deployment", service.getName()),
-                            DeploymentErrorCode.COMPONENT_BROKEN,
-                            DeploymentErrorCodeUtils.classifyComponentError(service, kernel));
-                }
-                if (!service.reachedDesiredState()) {
-                    allServicesRunning = false;
-                    continue;
-                }
-                if (State.RUNNING.equals(state) || State.FINISHED.equals(state) || !service.shouldAutoStart()
-                        && service.reachedDesiredState()) {
-                    continue;
-                }
-                allServicesRunning = false;
-            }
-            if (allServicesRunning) {
+        while (!isAllServiceInDesiredState(servicesToTrack, mergeTime, kernel)) {
+            Thread.sleep(WAIT_SVC_START_POLL_INTERVAL_MILLISEC); // hardcoded
+        }
+    }
+
+    /**
+     * Completes the provided future when all the listed services are running.
+     * Exits early if the future is cancelled
+     *
+     * @param servicesToTrack       services to track
+     * @param mergeTime             time the merge was started, used to check if a service is broken due to the merge
+     * @param kernel                kernel
+     * @param totallyCompleteFuture used to check if a deployment is cancelled
+     * @throws InterruptedException   if the thread is interrupted while waiting here
+     * @throws ServiceUpdateException if a service could not be updated
+     */
+    public static void waitForServicesToStart(Collection<GreengrassService> servicesToTrack, long mergeTime,
```

Comment:
It looks like you are not stopping ongoing lifecycle executions when you cancel a deployment. I am okay with that but we need to ensure 3 things
1. Nucleus should not be running a rogue component lifecycle thread
2. Nucleus should not block the execution of current deployment because a component lifecycle thread of a previously cancelled deployment is still running
3. Nucleus should always be running 1 version of any component at any point

I think we need to investigate these concerns a bit more

Reply:
Had offline discussion.

Components have a default timeout value of 120. As a result when a deployment is cancelled, by default the next deployment will be processed in 120 seconds.

For large timeouts deploying a component with bootstrap will restart the nucleus allowing users to unblock their device if a cancelled deployment remains stuck for too long.

Reply:
I am fine with stopping components as a bootstrap deployment to ensure rogue components stop eventually. I still want us to go in and fix component lifecycle to ensure no component lifecycle thread is running in the background. This can be a bug fix which is out of scope for this impl.